src/app/modules/auth/auth.controller.ts

```diff
@@ -8,167 +7,223 @@ import { createLoggerWithLabel } from '../../../config/logger'
+      })
   )
-  if (sendErr) {
-    logger.error({
-      message: 'Error mailing OTP',
-      meta: logMeta,
-      error: sendErr,
-    })
-
-    return res
-      .status(StatusCodes.INTERNAL_SERVER_ERROR)
-      .send(
-        'Error sending OTP. Please try again later and if the problem persists, contact us.',
-      )
-  }
-
-  // Successfully sent login otp.
-  logger.info({
-    message: 'Login OTP sent successfully',
-    meta: logMeta,
-  })
-
-  return res.status(StatusCodes.OK).send(`OTP sent to ${email}!`)
 }
 
 /**
- * Precondition: AuthMiddlewares.validateDomain must precede this handler.
+ *  Handler for POST /auth/verifyotp endpoint.
+ * @returns 200 when user has successfully logged in, with session cookie set
+ * @returns 401 when the email domain is invalid
+ * @returns 422 when the OTP is invalid
+ * @returns 500 when error occurred whilst verifying the OTP
  */
-export const handleLoginVerifyOtp: RequestHandler = async (
-  req: Request<
-    ParamsDictionary,
-    string | SessionUser,
-    { email: string; otp: string }
-  >,
-  res: ResponseAfter['validateDomain'],
-) => {
+export const handleLoginVerifyOtp: RequestHandler<
+  ParamsDictionary,
+  string | SessionUser,
+  { email: string; otp: string }
+> = async (req, res) => {
   // Joi validation ensures existence.
   const { email, otp } = req.body
-  // validateDomain middleware will populate agency.
-  const { agency } = res.locals
 
   const logMeta = {
     action: 'handleLoginVerifyOtp',
     email,
     ip: getRequestIp(req),
   }
 
-  const [verifyErr] = await to(AuthService.verifyLoginOtp(otp, email))
-
-  if (verifyErr) {
-    logger.warn({
-      message:
-        verifyErr instanceof ApplicationError
-          ? 'Login OTP is invalid'
-          : 'Error occurred when trying to validate login OTP',
+  const validateResult = await AuthService.validateEmailDomain(email)
+  if (validateResult.isErr()) {
+    const { error } = validateResult
+    logger.error({
+      message: 'Domain validation error',
       meta: logMeta,
-      error: verifyErr,
+      error,
     })
+    const { errorMessage, statusCode } = mapRouteError(error)
+    return res.status(statusCode).send(errorMessage)
+  }
 
-    if (verifyErr instanceof ApplicationError) {
-      return res.status(verifyErr.status).send(verifyErr.message)
-    }
+  // Since there is no error, agency is retrieved from validation.
+  const agency = validateResult.value
 
-    // Unknown error, return generic error response.
-    return res
-      .status(StatusCodes.INTERNAL_SERVER_ERROR)
-      .send(
-        'Failed to validate OTP. Please try again later and if the problem persists, contact us.',
-      )
-  }
+  return AuthService.verifyLoginOtp(otp, email)
+    .map(async () => {
+      // OTP is valid, proceed to login user.
+      try {
+        // TODO (#317): remove usage of non-null assertion
+        const user = await UserService.retrieveUser(email, agency)
```

Comment:
might it be worth it to convert this function to return a `ResultAsync` as well? since you've already made all this effort haha

Reply:
Done in 6171717 and 3c8da7d